cloud: Initial add of Prometheus / AlertManager configs.

bobvawter opened the pull request:
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26989)
<!-- Reviewable:end -->

jseldess commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/README.md, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0ccbpB9vDoX0Be6xJ:-LG0ccbpB9vDoX0Be6xK:b-3zksu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L19)):*
> ```Markdown
> 
> 
> If you're running on Google Kubernetes Engine, it's necessary to ensure
> ```

We call this out already as step 4 here for secure kubernetes deployment: https://www.cockroachlabs.com/docs/stable/orchestrate-cockroachdb-with-kubernetes.html#step-2-start-kubernetes

But this step is missing for insecure deployments. Is it now required in that case as well?

---

*[cloud/kubernetes/prometheus/README.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0d8qp4TxoqsaGmtx8:-LG0d8qp4TxoqsaGmtx9:bbed08z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L38)):*
> ```Markdown
> https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.20/bundle.yaml`
> 
> Ensure that prometheus-operator has started before continuing.
> ```

How do you ensure this? Is there a command that gives you that confirmation?

---

*[cloud/kubernetes/prometheus/README.md, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0dTih1TFVfYq9t5O4:-LG0dTih1TFVfYq9t5O5:bzgit7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L40)):*
> ```Markdown
> Ensure that prometheus-operator has started before continuing.
> 
> Create the various objects necessary to run a prometheus instance:
> ```

Running this as-is gives me an error. Do we need the full path to this yaml? 

```
~$ kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.20/bundle.yaml
clusterrolebinding "prometheus-operator" created
clusterrole "prometheus-operator" created
serviceaccount "prometheus-operator" created
deployment "prometheus-operator" created
~$ kubectl apply -f prometheus.yaml
error: the path "prometheus.yaml" does not exist
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/README.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0d8qp4TxoqsaGmtx8:-LG0gJ7c7_nStUy6Ksiq:btjtffs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L38)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

How do you ensure this? Is there a command that gives you that confirmation?
</blockquote></details>

`kubectl get deploy prometheus-operator` showing available==1

---

*[cloud/kubernetes/prometheus/README.md, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0dTih1TFVfYq9t5O4:-LG0fie25odC632hQyVp:b-s7zl6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L40)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Running this as-is gives me an error. Do we need the full path to this yaml? 

```
~$ kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.20/bundle.yaml
clusterrolebinding "prometheus-operator" created
clusterrole "prometheus-operator" created
serviceaccount "prometheus-operator" created
deployment "prometheus-operator" created
~$ kubectl apply -f prometheus.yaml
error: the path "prometheus.yaml" does not exist
```
</blockquote></details>

That's referring to the `prometheus.yaml` that's in the same directory.

For the moment, you can use https://raw.githubusercontent.com/bobvawter/cockroach/prom/cloud/kubernetes/prometheus/prometheus.yaml

I can update these references to a github link referencing the cockroach repo once this is checked in.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/README.md, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0ccbpB9vDoX0Be6xJ:-LG0gYdd3Suc6vBByPpV:b-n0dpr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L19)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

We call this out already as step 4 here for secure kubernetes deployment: https://www.cockroachlabs.com/docs/stable/orchestrate-cockroachdb-with-kubernetes.html#step-2-start-kubernetes

But this step is missing for insecure deployments. Is it now required in that case as well?
</blockquote></details>

It's required to grant the prometheus components the necessary permissions, even in an insecure cluster.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/README.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0d8qp4TxoqsaGmtx8:-LG0xO2k9uHTKJm5eAcT:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L38)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

`kubectl get deploy prometheus-operator` showing available==1
</blockquote></details>

Thanks.

---

*[cloud/kubernetes/prometheus/README.md, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0dTih1TFVfYq9t5O4:-LG0xM_eAZC0RonbOdlN:b-jhyysx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L40)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

That's referring to the `prometheus.yaml` that's in the same directory.

For the moment, you can use https://raw.githubusercontent.com/bobvawter/cockroach/prom/cloud/kubernetes/prometheus/prometheus.yaml

I can update these references to a github link referencing the cockroach repo once this is checked in.
</blockquote></details>

Got it. Thanks.

---

*[cloud/kubernetes/prometheus/README.md, line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0svJE1L8LERxEglmZ:-LG0svJE1L8LERxEglm_:bjzlr9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L46)):*
> ```Markdown
> * `kubectl port-forward
> prometheus-cockroachdb-0 9090`
> * Open http://localhost:9093 in your browser.
> ```

Should be `localhost:9090`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 10 of 10 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/alert-rules.yaml, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0u9DN5pBemERQXuWN:-LG0u9DO-8IzqgQveX-X:bj3dhns) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/alert-rules.yaml#L1)):*
> ```YAML
> apiVersion: monitoring.coreos.com/v1
> ```

I take it there's no way to reference or include our existing monitoring/rules/alerts.yml file instead of duplicating it here?

If not, are there any comments or documentation that we could add that would at least make it less likely they'd diverge over time?

---

*[cloud/kubernetes/prometheus/alert-rules.yaml, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0uHCLAQKfLhW0KRFQ:-LG0uHCLAQKfLhW0KRFR:b-h8nmo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/alert-rules.yaml#L4)):*
> ```YAML
> kind: PrometheusRule
> metadata:
>   creationTimestamp: null
> ```

Just delete this line.

---

*[cloud/kubernetes/prometheus/README.md, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0yp-c0Qg1zhoqiO7F:-LG0yp-c0Qg1zhoqiO7G:b-4qy69g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L5)):*
> ```Markdown
> 
> 
> References used:
> ```

Tiny nit, but I think it'd make more sense for the getting-started link to be listed above the alerting link.

---

*[cloud/kubernetes/prometheus/README.md, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0yz6c0q8NuGvN5HMS:-LG0yz6c0q8NuGvN5HMT:bpdjb0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L16)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> * `kubectl apply -f
> https://raw.githubusercontent.com/cockroachdb/cockroach/master/cloud/kubernetes/v1.7/cockroachdb-statefulset.yaml`
> * `kubectl apply -f
> https://raw.githubusercontent.com/cockroachdb/cockroach/master/cloud/kubernetes/v1.7/cluster-init.yaml`
>  
> ```
</details>


Why the v1.7 versions?

---

*[cloud/kubernetes/prometheus/README.md, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0z9MS8I8zwLbocXaX:-LG0z9MS8I8zwLbocXaY:b-ranin0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L20)):*
> ```Markdown
> 
> If you're running on Google Kubernetes Engine, it's necessary to ensure
> that your k8s user is part of the cluster-admin groups.  Edit the
> ```

This is more a question for @jseldess, but should we be using the abbreviation here? I'd just stick with spelling out Kubernetes.

---

*[cloud/kubernetes/prometheus/README.md, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0zQ7XAmN87iAONlhM:-LG0zQ7XAmN87iAONlhN:bsiq7bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L28)):*
> ```
> Edit cockroachdb service to add the label prometheus:cockroachdb We use
> ```

Edit the cockroachdb service to add the label `prometheus: cockroachdb`. We use

(missing "the", formatting on the label, and a period)

---

*[cloud/kubernetes/prometheus/README.md, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0z_ad0gaCAh9rEImX:-LG0z_aeB0sWTQ6WVyK8:bfz9vpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L29)):*
> ```Markdown
> 
> Edit cockroachdb service to add the label prometheus:cockroachdb We use
> this because we don't want to duplicate the monitoring data between two
> ```

You might want to expand on this a little more, since I'm not sure exactly what it means or why it's necessary from this one sentence.

---

*[cloud/kubernetes/prometheus/README.md, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0zjEg5XDf94x-3tb5:-LG0zjEg5XDf94x-3tb6:b-t4aoi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L36)):*
> ```Markdown
> Install Prometheus Operator:
> * `kubectl apply -f
> https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.20/bundle.yaml`
> ```

Is using a particular version rather than master intentional? What's your take on the tradeoff between their master potentially changing vs us having to update this in the future?

---

*[cloud/kubernetes/prometheus/README.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0d8qp4TxoqsaGmtx8:-LG0zuc0C3ZWG6bYVOo9:bjdcb4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L38)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Thanks.
</blockquote></details>

Should we include the command in the guide here? It's a reasonable question.

---

*[cloud/kubernetes/prometheus/README.md, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1-LmUDVv-noCeMkvO:-LG1-LmUDVv-noCeMkvP:b-c1psns) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L57)):*
> ```Markdown
> 
> Edit the template `alertmanager.yaml` with your relevant configuration.
> What's in the file has a dummy web hook, per the demo.
> ```

Per what demo?

---

*[cloud/kubernetes/prometheus/README.md, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1-g-F62I6Ato9oMl4:-LG1-g-F62I6Ato9oMl5:ba48lv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L68)):*
> ```Markdown
> Create an AlertManager object to run a replicated AlertManager instance
> and create a ClusterIP service so that Prometheus can forward alerts:
> * `kubectl apply -f alerts.yaml`
> ```

The file naming here is a little odd. alertmanager.yaml sounds like it would create the alert manager, but it doesn't. alerts.yaml sounds like it'd contain the alerting rules, but it doesn't. How about:

* Rename alertmanager.yaml to alertmanager-config.yaml
* Rename alerts.yaml to alertmanager.yaml

---

*[cloud/kubernetes/prometheus/README.md, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG10JuS6Rkeo6H6lsYl:-LG10JuS6Rkeo6H6lsYm:b-hfejbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L73)):*
> ```Markdown
> Verify that AlertManager is running:
> * `kubectl port-forward alertmanager-cockroachdb-0  9093`
> * Open http://localhost:9093 in your browser.
> ```

Maybe mention that it should look like the screenshot below? Otherwise I thought for a minute that the screenshot was supposed to be the prometheus page from the next bullet point

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[cloud/kubernetes/prometheus/README.md, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0z9MS8I8zwLbocXaX:-LG11l9w96qWIgUkg8bJ:b-1wpqcc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L20)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

This is more a question for @jseldess, but should we be using the abbreviation here? I'd just stick with spelling out Kubernetes.
</blockquote></details>

Good point. We don't use the k8s abbreviation anywhere in our docs or other configs, as far as I know, so I'd avoid that here as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

jseldess commented:
At the end of our Kubernetes tutorial, we explain how to remove all resources before stopping the cluster. How should we expand this to capture all of the prometheus and alertmanager resources as well? Include a different label?

```
kubectl delete pods,statefulsets,services,persistentvolumeclaims,persistentvolumes,poddisruptionbudget,jobs,rolebinding,clusterrolebinding,role,clusterrole,serviceaccount -l app=cockroachdb
```

Same question for secrets:

```
kubectl delete secrets default.client.root default.node.cockroachdb-0 default.node.cockroachdb-1 default.node.cockroachdb-2 default.node.cockroachdb-3
```

Would be great if you could include this guidance toward the end of this guide.

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG14QWo53vJGCufFbh2:b97q4t0)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
I have this working now with a secure cluster configuration.  One thing that is currently broken is that all of the existing alertmanager rules use `job=cockroach`, but the metrics are actually stored in prometheus as `job=cockroach*db*`.  Which is right?

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[cloud/kubernetes/prometheus/alert-rules.yaml, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0u9DN5pBemERQXuWN:-LG1dj2A1bMNO6IDG3wR:bnebotu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/alert-rules.yaml#L1)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I take it there's no way to reference or include our existing monitoring/rules/alerts.yml file instead of duplicating it here?

If not, are there any comments or documentation that we could add that would at least make it less likely they'd diverge over time?
</blockquote></details>

The operator overwrites the `rule_files` configuration; it's not possible to add additional files to the secret that contains the `alertmanager.yaml` config file and have them be referenced.  How about a shell script that, given an existing yaml file, wraps it with the right prelude to make it a proper `PrometheusRule` object?

---

*[cloud/kubernetes/prometheus/alert-rules.yaml, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0uHCLAQKfLhW0KRFQ:-LG1eGHD0nHPcfdWcYNp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/alert-rules.yaml#L4)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Just delete this line.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0yp-c0Qg1zhoqiO7F:-LG1ewk_039iuu_8Gad-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L5)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Tiny nit, but I think it'd make more sense for the getting-started link to be listed above the alerting link.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0yz6c0q8NuGvN5HMS:-LG1eyNv29lf-b4WQGh6:beuwpf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L16)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> * `kubectl apply -f
> https://raw.githubusercontent.com/cockroachdb/cockroach/master/cloud/kubernetes/v1.7/cockroachdb-statefulset.yaml`
> * `kubectl apply -f
> https://raw.githubusercontent.com/cockroachdb/cockroach/master/cloud/kubernetes/v1.7/cluster-init.yaml`
>  
> ```

Why the v1.7 versions?
</blockquote></details>

Fixed.  Old copy-and-paste.

---

*[cloud/kubernetes/prometheus/README.md, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0z9MS8I8zwLbocXaX:-LG1f3_yEK0fIXJHD2I8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L20)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Good point. We don't use the k8s abbreviation anywhere in our docs or other configs, as far as I know, so I'd avoid that here as well.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0zQ7XAmN87iAONlhM:-LG1fCAS-EnnJfXaZSE_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L28)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> Edit cockroachdb service to add the label prometheus:cockroachdb We use
> ```

Edit the cockroachdb service to add the label `prometheus: cockroachdb`. We use

(missing "the", formatting on the label, and a period)
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0z_ad0gaCAh9rEImX:-LG1fTve1ZW4ra4xZ3tP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L29)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You might want to expand on this a little more, since I'm not sure exactly what it means or why it's necessary from this one sentence.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0zjEg5XDf94x-3tb5:-LG1fUxq36sD-mbTVZHz:b-6mt4uk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L36)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Is using a particular version rather than master intentional? What's your take on the tradeoff between their master potentially changing vs us having to update this in the future?
</blockquote></details>

Their master actually did change in an incompatible way.  The `PrometheusRule` object was very recently named `RuleFile`.  I get the impression that the operator's API is still evolving.  For a quick-start, I'd rather give someone an outdated version of the software and a known-good config.

---

*[cloud/kubernetes/prometheus/README.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0d8qp4TxoqsaGmtx8:-LG1g94I5mfDb1TZ4uWS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L38)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should we include the command in the guide here? It's a reasonable question.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0svJE1L8LERxEglmZ:-LG1gCNf8noudcEHO2jM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L46)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Should be `localhost:9090`.
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1-LmUDVv-noCeMkvO:-LG1gIpQ8SMZ_fBDe1mZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L57)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Per what demo?
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1-g-F62I6Ato9oMl4:-LG1gJZs876-MRjHohA5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L68)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

The file naming here is a little odd. alertmanager.yaml sounds like it would create the alert manager, but it doesn't. alerts.yaml sounds like it'd contain the alerting rules, but it doesn't. How about:

* Rename alertmanager.yaml to alertmanager-config.yaml
* Rename alerts.yaml to alertmanager.yaml
</blockquote></details>

Done.

---

*[cloud/kubernetes/prometheus/README.md, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG10JuS6Rkeo6H6lsYl:-LG1ge0QE7p3YE20Wiwh:b-mavatn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L73)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Maybe mention that it should look like the screenshot below? Otherwise I thought for a minute that the screenshot was supposed to be the prometheus page from the next bullet point
</blockquote></details>

Reworded to be more clear.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG1qQj7ASO6etxZC7bF:bqbg179)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Very nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> other than my question about the new labels and your `cockroach` vs `cockroachdb` question. I'd think cockroachdb, but @mberhault can set the standard.

---

Reviewed 6 of 7 files at r2.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[cloud/kubernetes/prometheus/alert-rules.yaml, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0u9DN5pBemERQXuWN:-LG1wCjDE4-5OhsH9NfB:b-skw559) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/alert-rules.yaml#L1)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

The operator overwrites the `rule_files` configuration; it's not possible to add additional files to the secret that contains the `alertmanager.yaml` config file and have them be referenced.  How about a shell script that, given an existing yaml file, wraps it with the right prelude to make it a proper `PrometheusRule` object?
</blockquote></details>

If you want to write that, that'd work. I'd also be fine just sticking a note into a README.md in the monitoring/rules directory.

---

*[cloud/kubernetes/prometheus/alertmanager.yaml, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1wa-55KFFtIGvpJjr:-LG1wa-55KFFtIGvpJjs:byvnt43) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/alertmanager.yaml#L7)):*
> ```YAML
>   name: cockroachdb
>   labels:
>     app: cockroachdb
> ```

Does using `app: cockroachdb` here and in the selector below (and in prometheus.yaml) not mess with the services that we set up for cockroachdb itself or with the statefulset's tracking of how many pods are currently running? I'd think that using the same label would cause problems.

---

*[cloud/kubernetes/prometheus/README.md, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG0zjEg5XDf94x-3tb5:-LG1xABmAJMPVGyg83sq:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba6d7c2ebda136b81aecdecafaf0554462da08/cloud/kubernetes/prometheus/README.md#L36)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Their master actually did change in an incompatible way.  The `PrometheusRule` object was very recently named `RuleFile`.  I get the impression that the operator's API is still evolving.  For a quick-start, I'd rather give someone an outdated version of the software and a known-good config.
</blockquote></details>

:thumbsup: 

---

*[cloud/kubernetes/prometheus/README.md, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1xNXEEt6r6z1pcQJk:-LG1xNXEEt6r6z1pcQJl:bk6h0lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/README.md#L70)):*
> ```Markdown
> alerting guide linked from the top of the document.
> 
> Upload `alertmanager-config.yaml`, renaming it to `alertmanager.yaml`
> ```

Ah, I guess that was the reason for the naming. I didn't realize, but thanks for making it work.

---

*[cloud/kubernetes/prometheus/README.md, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1xUMMBosIhyW2DJNq:-LG1xUMMBosIhyW2DJNr:bancp5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/README.md#L86)):*
> ```Markdown
> Verify that AlertManager is running:
> * `kubectl port-forward alertmanager-cockroachdb-0  9093`
> * Open http://localhost:9093 in your browser.  You shoud see something
> ```

s/shoud/should/

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG1x_2z6mUQBSv1Ume5:b-684ise)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Nice!

However, it seems the label is set to `job=cockroachdb`. The alerts all use `job=cockroach` to ensure only the desired metrics are used. This will make alerts never trigger.

bobvawter commented:
@mberhault, the use of `cockroach` in the alert configs vs `cockroachdb` in the k8s configs is the crux of my question.  It would seem weird to have a from-zero quickstart flow that's inconsistent in its naming, but maintaining two sets of alert configs also seems problematic.  Unless you have a better suggestion, I'm going to write a little utility that will wrap the existing alert configs in the necessary yaml to create a `PrometheusRule` object and also do an `s/cockroach/cockroachdb/g`.

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG5Q5Ia9QOj2hEqJvCM:by1xydw)*
<!-- Sent from Reviewable.io -->

mberhault commented:
The easiest might be to rename our existing alert rules to use `job=cockroachdb`. We'll have to either relabel the existing metrics or just live with broken alerts for the transition (should be ok, most of our alerts are over a short interval anyway). It will be a one time annoyance but allows us to be consistent between configs as well as our use of `cockroachdb` instead of `cockroach`.

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG5RAfqBs0G2mRyGqsm:bdehjb7)*
<!-- Sent from Reviewable.io -->

bobvawter committed:
cloud: Initial add of Prometheus / AlertManager configs.

The main monitoring rules have been tweaked to change from job=cockroach to
job=cockroachdb to align to the naming convention used in our k8s quickstart
configurations.

A new wraprules tool has been added to combine the reference prometheus rules
into a single PrometheusRule configuration object.

Release note: None

bobvawter commented:
PTAL: I've added a tool to roll up our reference configs into the `alert-rules.yaml` file and renamed the monitoring to use `job=cockroachdb`.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG60KCK5za2D839WSKx:b-mlyimq)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[cloud/kubernetes/prometheus/alertmanager.yaml, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1wa-55KFFtIGvpJjr:-LG619rpB3M0sGeVXLvN:b-5nojwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/alertmanager.yaml#L7)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Does using `app: cockroachdb` here and in the selector below (and in prometheus.yaml) not mess with the services that we set up for cockroachdb itself or with the statefulset's tracking of how many pods are currently running? I'd think that using the same label would cause problems.
</blockquote></details>

The actual pods that are derived from this `Alertmanager` are generated with `app=alertmanager`.  I'm just using this label to make it easy to write a cleanup rule.

---

*[cloud/kubernetes/prometheus/README.md, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1xUMMBosIhyW2DJNq:-LG61Tjn10SXxcc5kxG0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/README.md#L86)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/shoud/should/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[cloud/kubernetes/prometheus/alertmanager.yaml, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-LG1wa-55KFFtIGvpJjr:-LG62eIM-2Jh2E1-o5hk:b-qm0anp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8c688b9485d8c8dacf370bfbab13b91a96f4b4/cloud/kubernetes/prometheus/alertmanager.yaml#L7)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

The actual pods that are derived from this `Alertmanager` are generated with `app=alertmanager`.  I'm just using this label to make it easy to write a cleanup rule.
</blockquote></details>

Thanks for clarifying, I was concerned that the labels got passed along to the pods managed by the operator.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26989#-:-LG63CPWEKzPU-fxJICl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@bobvawter, should I try to run through the instructions with a secure cluster again?

bobvawter commented:
@jseldess Yes, please. It would be good to do another full-up test to make sure nothing has drifted.

bobvawter commented:
bors r+

bobvawter commented:
bors retry

bobvawter commented:
bors r+
